### PR TITLE
Allow mutual funding of a new channel

### PR DESCRIPTION
**Pull request by gertjaap**
This pull request adds a new command to lit: ```dualfund```

The command allows peers to fund a channel together, where both sides commit money into the funding TXO. This is particularly useful when we are moving ahead with the Discreet Log Contracts, where mutual funding is kind of mandatory.

While dualfunding, the initiator will specify the amount he wishes to fund, and the amount he requests the peer to fund. The peer will receive this as a pending request he can accept or decline. Once accepted, the channel will be created (as before) and funded (using inputs from both peers).

The command has three subcommands: ```start```, ```accept``` and ```decline```

When connected to a peer, you can issue the start command like so:

```dualfund start [peer index] [coin type] [our amount] [their amount]```

For instance: ```dualfund start 1 257 500000 500000```

When issuing the command, we will ask the remote peer to fund 500k satoshi, while we fund that amount as well. The amounts are freely adjustable, as long as the total is 1M satoshi or higher (the existing funding limit)

When receiving a request to dual fund, you can review the funding request using ```ls``` and use the ```dualfund accept``` or ```dualfund decline``` commands to respond.

The sequence of communication is:

```
A -> B      Request funding UTXOs and change address for TheirAmount, 
               sending along our own for OurAmount

B -> A      Respond with UTXOs and change address for TheirAmount (Accept)
(or) B -> A Respond with a decline

A and B     Compose the funding TX

Then (regular funding messages):
A -> B      Requests channel point and refund pubkey, sending along its own

            A channel point (33) (channel pubkey for now)
            A refund (33)

B -> A      Replies with channel point and refund pubkey

            B channel point (32) (channel pubkey for now)
            B refund (33)

A -> B      Sends channel Description:
            ---
            outpoint (36)
            capacity (8)
            initial push (8)
            B's HAKD pub #1 (33)
            signature (~70)
            ---

Then (dual funding specific, signatures for the funding TX added to Ack):

B -> A      Acknowledges Channel:
            ---
            A's HAKD pub #1 (33)
			signature (~70)
			funding signatures
            ---

A           Publishes funding TX

=== time passes, fund tx gets in a block ===

A -> B      SigProof
            SPV proof of the outpoint (block height, tree depth, tx index, hashes)
            signature (~70)
```

### CODE REVIEW

**Comment by Varunram** [comment]
Unrelated to the PR (which is awesome!), but I wonder why Travis didn't run the test suite against the commits..

**Comment by alindr** [comment]
6/19/18 meeting - Gert-Jaap to make sure this can be merged.

**Comment by gertjaap** [comment]
Rebased this to master and tested it - channel is formed properly with funds on either side. Ready to merge.

**Comment by delbonis** [comment]
```
10:52:07 ../../qln/dualfund.go:378:25: cannot use &outPoint (type *"github.com/mit-dci/lit/vendor/github.com/adiabat/btcd/wire".OutPoint) as type *"github.com/mit-dci/lit/wire".OutPoint in assignment
10:52:07 ../../qln/dualfund.go:446:27: cannot use &u.Outpoint (type *"github.com/mit-dci/lit/wire".OutPoint) as type *"github.com/mit-dci/lit/vendor/github.com/adiabat/btcd/wire".OutPoint in argument to "github.com/mit-dci/lit/vendor/github.com/adiabat/btcd/wire".NewTxIn
10:52:07 ../../qln/dualfund.go:450:27: cannot use &u.Outpoint (type *"github.com/mit-dci/lit/wire".OutPoint) as type *"github.com/mit-dci/lit/vendor/github.com/adiabat/btcd/wire".OutPoint in argument to "github.com/mit-dci/lit/vendor/github.com/adiabat/btcd/wire".NewTxIn
10:52:07 ../../qln/dualfund.go:469:13: cannot use txo (type *"github.com/mit-dci/lit/wire".TxOut) as type *"github.com/mit-dci/lit/vendor/github.com/adiabat/btcd/wire".TxOut in argument to tx.AddTxOut
10:52:07 ../../qln/dualfund.go:617:18: cannot use fundingTx (type *"github.com/mit-dci/lit/vendor/github.com/adiabat/btcd/wire".MsgTx) as type *"github.com/mit-dci/lit/wire".MsgTx in argument to wal.SignMyInputs
10:52:07 ../../qln/dualfund.go:620:11: cannot use fundingTx (type *"github.com/mit-dci/lit/vendor/github.com/adiabat/btcd/wire".MsgTx) as type *"github.com/mit-dci/lit/wire".MsgTx in argument to lnutil.NewDualFundingChanAckMsg
10:52:07 ../../qln/dualfund.go:681:44: cannot use tx (type *"github.com/mit-dci/lit/vendor/github.com/adiabat/btcd/wire".MsgTx) as type *"github.com/mit-dci/lit/wire".MsgTx in argument to nd.SubWallet[qc.Coin()].SignMyInputs
10:52:07 ../../qln/dualfund.go:690:23: cannot use msg.SignedFundingTx.TxIn[i].Witness (type "github.com/mit-dci/lit/wire".TxWitness) as type "github.com/mit-dci/lit/vendor/github.com/adiabat/btcd/wire".TxWitness in assignment
10:52:07 ../../qln/dualfund.go:695:38: cannot use tx (type *"github.com/mit-dci/lit/vendor/github.com/adiabat/btcd/wire".MsgTx) as type *"github.com/mit-dci/lit/wire".MsgTx in argument to nd.SubWallet[qc.Coin()].DirectSendTx
```

The build failed, looks like there's some imports in `dualfund.go` that need to be updated.  Just swap out the `adiabat/btcd` imports to the merged ones.

File: cmd/lit-af/chancmds.go [hunk 1]
```diff
@@ -148,6 +180,102 @@ func (lc *litAfClient) FundChannel(textArgs []string) error {
...
+		fmt.Fprintf(color.Output, dualFundCommand.Description)
+		return nil
+	} else {
+		return fmt.Errorf(dualFundCommand.Format)
+	}
+}
+
+// Mutually fund a channel
+func (lc *litAfClient) DualFundChannel(textArgs []string) error {
+	if len(textArgs) > 0 && textArgs[0] == "-h" {
```
Comment (Varunram) [inline cmd/lit-af/chancmds.go]: We added a couple of handlers to reduce these calls, might be nice to have it over here after looking into the imports issue trey mentioned :)

**Comment by gertjaap** [comment]
Will take care of the mentioned issues by @delbonis and @Varunram and request merge again. With the rate of work being done on lit atm, making larger pull requests is working against a moving target.

**Comment by gertjaap** [comment]
I implemented the usage of `CheckHelpCommand` utility function. But there's an issue in that, i outlined in #231 .

This PR is ready to merge in my perspective.

**Comment by adiabat** [comment]
This is great, and how most people think lightning channels always work :)

**Pull request merged**